array/array.go

```diff
@@ -0,0 +1,131 @@
+package array
+
+import (
+	"errors"
+)
+
+type Array struct {
+	arr  []int
+	size int
+}
+
+func NewArray(capacity int) *Array {
+	return &Array{
+		arr: make([]int, capacity),
+	}
+}
+
+func (a *Array) Size() int {
+	return a.size
+}
+
+func (a *Array) Capacity() int {
+	return len(a.arr)
+}
+
+func (a *Array) IsEmpty() bool {
+	return a.Size() == 0
+}
+
+func (a *Array) ValueAt(index int) (int, error) {
+	if index < 0 || index >= a.Size() {
+		return -1, errors.New("Index is out of range")
+	}
+	return a.arr[index], nil
+}
+
+func (a *Array) Insert(item int, index int) error {
+	if index < 0 || index >= a.Capacity() {
+		return errors.New("Index is out of range")
+	}
+
+	a.checkAndShrink()
+
+	// shift array to the right
+	for i := a.size; i > index; i-- {
+		a.arr[i] = a.arr[i-1]
+	}
+
+	a.arr[index] = item
+
+	// insert at index out of size range
+	// will increase the size to index + 1
+	if index >= a.Size() {
+		a.size = index + 1
+	} else {
+		a.size++
+	}
+
+	return nil
+}
+
+func (a *Array) Append(item int) {
+	a.checkAndShrink()
+
+	a.arr[a.size] = item
+	a.size++
+}
+
+func (a *Array) Pop() (int, error) {
+	if a.IsEmpty() {
+		return -1, errors.New("Array is empty")
+	}
+
+	a.checkAndReduce()
+
+	a.size--
+	value := a.arr[a.size]
```

Comment:
At this point, `a.arr[a.size]` is an invalid value. I prefer to getting `value` before decreasing `size`.

Reply:
I will not fail, because I decrease the size before accessing.
Before I used to write like this:
```go
value := a.arr[a.size - 1]
a.size--
```
I just think the language has to do 2 subtract operations here (`a.size - 1` and `a.size--`), so I do the subtraction first then access the value later.

Reply:
It will not fail but values which are not in range [0, size] are invalid because we don't manage them. If you still want to decrease size before getting the value, you can `return a.arr[--a.size]` if it's available or at least don't create `value` variable.

Reply:
Well, `--a.size` in Go throws an error :(

For the `value` part, as I said, I try to minic JS implementation so when you delete an item at an index, you get the value in that index as returned value.

Reply:
Can we do it without creating `value` variable? How about `return arr[size], nil`?

Reply:
But the item will be lost, because I shift the array to the left right below that.

Reply:
I think you are misunderstanding. This is `pop` function and you don't need to shift anything here. There is only one more line in this function after this line `return value, nil`. Maybe you can check again.

Reply:
I must have been copy-pasted from the Remove function here. :( I see the mistake.